### DISCUSSION TITLE
Amazing Schwiizerdütsch

### DISCUSSION THREAD

**Discussion by esthicodes** [Estheryu991/SwissGerman_Dictionary#177] [General]
Reserviere en Parkplatz vo spezifischer Grössi: kompakt / regulär / gross
Vergib jedem Auto en einzigartige Zuweisig (2 Autos dörfed ned de gliche Parkplatz ha)
Wenn en Fahrzüg dä Parkplatz verlässt, berechne dä Preis basierend uf Grössi und Dauer vom Parkiere.
Stell d Rechnig für dä Parkplatz vor dä Abfahrt vom Auto.

![B7E309C2-A0B3-4149-94F5-A1BD9A0FC04B](https://user-images.githubusercontent.com/78131082/219266853-27c671cc-bd13-48fd-80b6-2b9d979147a6.jpeg)

**Comment by omparaday** [comment]
This editor is synced in real time with your peer.

Use it to share thoughts and resources, such as:
- Features scope
- API design
- Pseudo code for specific components
- Data model/schema
- Back-of-the-envelope calculations
- Reference links
- Link to whiteboard or diagram such as https://sketchboard.me/new

Good luck!


esthi, korea. 
okay. 
https://github.com/esthicodes
https://github.com/esthicodes/Awesome-Swiss-German/discussions/177


are you able to hear me?

Requirements:
1. Reserve a parking slot of specific size compact/regular/large
2. Allotment should be unique (2 cars should not be alloted same slot)
3. When vehicle moves out, charge the car based on size and time duration of parking
4. Collect bill for car before moving out.

Non functional requirements:
1. Parking slots are limited.
2. There can be more than 1 entry and 1 exit (n entry gates and m exit gates).
3. Slot allotments should be mutually exclusive.
4. Service should have high availability.
5. Service has to be highly consistent.
6. There are 5000 parking slots.
7. There are 10 entry gates and 10 exit gates.
8. When there is no slot available return "PARKING FULL".
9. Per second 1 car is accessing on average.

Database design:
1. SQL database for slots.
Schema:
slotid, type(compact/regular/large), available(boolean), car number, lastupdated

Bills table SQL
2. Bill number, car number, slotid, duration, bill amount, generated timestamp, transaction id,

Transaction table:
transaction id, bill number, state(success /failure)


High level design:
shall I draw in drawing board?
https://docs.google.com/drawings/d/1LXuIWecbzExJe-nzTyn7EEDN_ONypyN_ci3exZBndL8/edit?usp=sharing



APIs
GateManager:
1. More than 1 instance running.
2. It will handle the entry and exit of cars.
3. For both entry and exit it will contact allotment service for alloting or releasing a lot based on car type.
4. If PARKING FULL redirect cars to exit.
5. If alloted redirect cars to alloted slot.
6. If car is leaving slot, contact the billing service and generate bill(from billing service) and collect pay(via payment service).


RPCs
Allotment service:
1. Handle the queue of requests for entry and exit coming from GateManager.
2. Process the requests serially in order to avoid multiple cars alloted same slot.
3. Update the sql table parking slots with car id and time of entry.
4. Update the sql table parking slots with exit and time of exit. return the exit time to GateManager for billing.

select slotid from parking_table where slottype = <type> AND available=true
update parking_table set carnumber=<no>, 

DatabaseManager:
1. Periodically backup the parking slot table (once every 5 mins).
2. IF parking slot table is down, redirect requests to backup table.
3. Peridically move the bills to low cost database or storage objects for cost saving (once per day).

Billing Service:
1. Handle bill generation requests from Gate Manager.
2. Generate bill and update in database based on duration and slot type.
3. Return the bill to gatemanager.

Payment Service:
1. Handle payments success/failure and keep records of payments
2. Update billing table with success trasactions.
3. Update trasactions table with success and failure.

Summary:



I completed.
Thank you :)

ok

amazing
